### ISSUE TITLE
Segmentation fault with enabled apcu extension

### ISSUE THREAD

**Issue by foment** [composer/composer#7436]
Hello folks,

last week I've tried to install the dependencies for my project using composer and got this weird PHP Fatal error, when using the latest composer version (Snapshot), downloaded directly from https://getcomposer.org/composer.phar.

When I run this command:

```
[vagrant@vbox /vagrant/www]$ php -d apc.enable_cli=1 composer.phar update
```

I get the following output:

```
PHP Fatal error:  PHP Startup: apc_mmap: mkstemp on /tmp/apc.A9HEJu failed: in Unknown on line 0
```

OS: CentOS 6.9 running in vagrant
PHP 7.2.3
APCu 5.1.11

In php.ini, apcu extension with cli support is of course enabled. When I disable APCu everything is working fine.
As a fix, I tried to change permissions of /tmp directory to 777, or to use another directory for temp files, but it didn't help.

As a note, after executing php composer.phar install command, it fails immediately, so I think that composer.json isn't even parsed, nor this issue has anything to do with it.

Here is also additional info from system log:

Core was generated by `/usr/bin/php -c /tmp/WCG8NT composer.phar update'.
Program terminated with signal 11, Segmentation fault.
#0  0x00007fe28bcd8a08 in pthread_rwlock_init () from /lib64/libpthread.so.0

When using latest stable version of composer (for example 1.6.5 https://getcomposer.org/download/1.6.5/composer.phar), everything is working fine, even when apcu is enabled.

Cheers

**Comment by johnstevenson** [comment]
If `xdebug` is involved, it could be related to https://github.com/composer/xdebug-handler/issues/57

**Comment by foment** [comment]
It look like you are right.
After disabling xdebug, composer update worked fine.

As additional info, if needed, host machine is Win10.

**Comment by johnstevenson** [comment]
Cool. Could you show me your Vagrant file, please, plus the apcu ini settings that you are using, and anything else so I can reproduce this on Win 10. Thanks.

**Comment by foment** [comment]
Here are the apcu ini settings:

> apc.smart=0
; The mktemp-style file_mask to pass to the mmap module
apc.mmap_file_mask=/tmp/apc.XXXXXX
; If enabled, APCu attempts to prevent "slamming" of a key.
apc.slam_defense=1
; Defines which serializer should be used. Default is the standard PHP serializer.
apc.serializer='default'
; Setting this enables APC for the CLI version of PHP (Mostly for testing and debugging).
apc.enable_cli=1
; RFC1867 File Upload Progress hook handler
apc.rfc1867=0
apc.rfc1867_prefix =upload_
apc.rfc1867_name=APC_UPLOAD_PROGRESS
apc.rfc1867_freq=0
apc.rfc1867_ttl=3600
; Enables APC handling of signals, such as SIGSEGV, that write core files when signaled.
; APC will attempt to unmap the shared memory segment in order to exclude it from the core file
apc.coredump_unmap=0

> 

Unfortunately, I am not allowed to upload Vagrantfile, as there are lots of company related things in it and it is quite big. However, I could tell you that we had no problems with the same Vagrantfile using CentOS 6.8, but it started happening with the box upgrade to the CentOS 6.9

**Comment by Dutchy-** [comment]
We have the same problem running Composer >1.7 on docker/CentOS. Running with either `apc.enable_cli=0` or xdebug disabled works:

Fails:
```
# php -d apc.enable_cli=1 /usr/bin/composer --help
PHP Fatal error:  PHP Startup: apc_mmap: mkstemp on /tmp/apc.hb3Wnx failed: in Unknown on line 0
```

Works:
```
# php -d apc.enable_cli=0 /usr/bin/composer --help
Do not run Composer as root/super user! See https://getcomposer.org/root for details
Usage:
  help [options] [--] [<command_name>]
```

To disable xdebug we start php with a minimum of configuration using `-n`:
```
# php -n -d extension_dir=/opt/remi/php72/root/usr/lib64/php/modules -d extension=phar.so -d extension=json.so -d extension=mbstring.so /usr/bin/composer --help
Usage:
  help [options] [--] [<command_name>]
```

This all worked fine on Composer 1.6.5 but stopped working for version 1.7.0 up to 1.7.2 at the time of writing.

Additional information:

```
# php --version
PHP 7.2.9 (cli) (built: Aug 15 2018 08:05:45) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.9, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v2.6.1, Copyright (c) 2002-2018, by Derick Rethans
```

```
# php --info | grep apc
/etc/opt/remi/php72/php.d/40-apcu.ini,
apcu
MMAP File Mask => /tmp/apc.dmrrwh
apc.coredump_unmap => Off => Off
apc.enable_cli => On => On
apc.enabled => On => On
apc.entries_hint => 4096 => 4096
apc.gc_ttl => 3600 => 3600
apc.mmap_file_mask => /tmp/apc.dmrrwh => /tmp/apc.dmrrwh
apc.preload_path => no value => no value
apc.serializer => igbinary => igbinary
apc.shm_segments => 1 => 1
apc.shm_size => 32M => 32M
apc.slam_defense => On => On
apc.smart => 0 => 0
apc.ttl => 0 => 0
apc.use_request_time => On => On
apc.writable => /tmp => /tmp
```

```
# php --info | grep xdebug
/etc/opt/remi/php72/php.d/15-xdebug.ini,
xdebug
xdebug support => enabled
xdebug.auto_trace => Off => Off
xdebug.cli_color => 0 => 0
xdebug.collect_assignments => Off => Off
xdebug.collect_includes => On => On
xdebug.collect_params => 0 => 0
xdebug.collect_return => Off => Off
xdebug.collect_vars => Off => Off
xdebug.coverage_enable => On => On
xdebug.default_enable => On => On
xdebug.dump.COOKIE => no value => no value
xdebug.dump.ENV => no value => no value
xdebug.dump.FILES => no value => no value
xdebug.dump.GET => no value => no value
xdebug.dump.POST => no value => no value
xdebug.dump.REQUEST => no value => no value
xdebug.dump.SERVER => no value => no value
xdebug.dump.SESSION => no value => no value
xdebug.dump_globals => On => On
xdebug.dump_once => On => On
xdebug.dump_undefined => Off => Off
xdebug.extended_info => On => On
xdebug.file_link_format => no value => no value
xdebug.filename_format => no value => no value
xdebug.force_display_errors => Off => Off
xdebug.force_error_reporting => 0 => 0
xdebug.gc_stats_enable => Off => Off
xdebug.gc_stats_output_dir => /tmp => /tmp
xdebug.gc_stats_output_name => gcstats.%p => gcstats.%p
xdebug.halt_level => 0 => 0
xdebug.idekey => no value => no value
xdebug.max_nesting_level => 256 => 256
xdebug.max_stack_frames => -1 => -1
xdebug.overload_var_dump => 2 => 2
xdebug.profiler_aggregate => Off => Off
xdebug.profiler_append => Off => Off
xdebug.profiler_enable => Off => Off
xdebug.profiler_enable_trigger => Off => Off
xdebug.profiler_enable_trigger_value => no value => no value
xdebug.profiler_output_dir => /tmp => /tmp
xdebug.profiler_output_name => cachegrind.out.%p => cachegrind.out.%p
xdebug.remote_addr_header => no value => no value
xdebug.remote_autostart => Off => Off
xdebug.remote_connect_back => Off => Off
xdebug.remote_cookie_expire_time => 3600 => 3600
xdebug.remote_enable => Off => Off
xdebug.remote_handler => dbgp => dbgp
xdebug.remote_host => localhost => localhost
xdebug.remote_log => no value => no value
xdebug.remote_mode => req => req
xdebug.remote_port => 9000 => 9000
xdebug.remote_timeout => 200 => 200
xdebug.scream => Off => Off
xdebug.show_error_trace => Off => Off
xdebug.show_exception_trace => Off => Off
xdebug.show_local_vars => Off => Off
xdebug.show_mem_delta => Off => Off
xdebug.trace_enable_trigger => Off => Off
xdebug.trace_enable_trigger_value => no value => no value
xdebug.trace_format => 0 => 0
xdebug.trace_options => 0 => 0
xdebug.trace_output_dir => /tmp => /tmp
xdebug.trace_output_name => trace.%c => trace.%c
xdebug.var_display_max_children => 128 => 128
xdebug.var_display_max_data => 512 => 512
xdebug.var_display_max_depth => 3 => 3
```

This happens both when running docker on my Mac but also in our Gitlab CI system running Linux. A common factor I see here is CentOS but we run a higher version:

```
# cat /etc/centos-release
CentOS Linux release 7.5.1804 (Core)
```

**Comment by johnstevenson** [comment]
@Dutchy- Thanks for that info. Have you got a docker file that I can reproduce this with please?

**Comment by dtakken** [comment]
Hello,

Same problem here on Docker / CentOS. The following Docker file reproduces the problem:

```
FROM centos:7

RUN yum install -y wget && \
    wget -q -O /etc/yum.repos.d/epel7.repo https://mirror.openshift.com/mirror/epel/epel7.repo && \
    yum-config-manager --enable epel

RUN wget http://rpms.famillecollet.com/enterprise/remi-release-7.rpm && \
    rpm -Uvh --nodeps remi-release-7*.rpm && \

    yum install -y --enablerepo=remi \
      php72-php-xml \
      php72-php-cli \
      php72-php-mbstring \
      php72-php-pecl-apcu \
      php72-php-opcache \
      php72-php-pecl-xdebug && \
      ln -s /bin/php72 /bin/php

RUN php -d allow_url_fopen=1 -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" && \
    php -d allow_url_fopen=1 composer-setup.php --install-dir=/usr/bin --filename=composer

RUN php --info
RUN php -d apc.enable_cli=1 /usr/bin/composer --help

```
Just build it using ```cat Dockerfile | docker build -t testcase -``` and it prints

```PHP Fatal error:  PHP Startup: apc_mmap: mkstemp on /tmp/apc.TuDVyX failed: in Unknown on line 0
The command '/bin/sh -c php -d apc.enable_cli=1 /usr/bin/composer --help' returned a non-zero code: 139
```

while building. Thanks for looking into it!

**Comment by johnstevenson** [comment]
Okay, this happens because xdebug-handler now merges the current ini settings into the ini files content (introduced in composer 1.70). It does this to pick up command-line ini-settings.

However, the `apc.mmap_file_mask` template value gets replaced with the name of the file being used, causing `mkstemp` to fail, presumably because the template value passed to it doesn't end `XXXXXX`.

At least it is an easy fix.

**Comment by johnstevenson** [comment]
@Seldaek This has been fixed in [xdebug-handler 1.2.1](https://github.com/composer/xdebug-handler/releases/tag/1.2.1)

**Comment by dtakken** [comment]
No problem @johnstevenson. Thanks a lot for fixing!

**Comment by michaelbutler** [comment]
Could you provide more insight on how to fix this error? I have the latest version of `composer` (1.7.2) under `vendor/bin/composer` and I tried installing `xdebug-handler` 1.3 but I'm still getting the segmentation fault.  Does `composer` itself include xdebug-handler in its phar? If not, this would be a chicken-and-egg problem, where `composer` cannot install the `xdebug-handler` code with the fix because it seg faults.

**Comment by johnstevenson** [comment]
@michaelbutler Yes Composer includes xdebug-handler but the fix isn't in the latest release (1.7.2).

Until Composer gets updated, you can either add `apc.enable_cli=0` to your php.ini or run with xdebug via `COMPOSER_ALLOW_XDEBUG=1`

**Comment by michaelbutler** [comment]
Thanks a lot, just realized after posting that the update 27 days ago was after 1.7.2. 👍
